File: R/303-range-sum-query-immutable/main.go
package main

import "fmt"

type NumArray struct {
	Sum []int
}

func Constructor(nums []int) NumArray {
	sum := make([]int, len(nums))
	s := 0
	for i := 0; i < len(nums); i++ {
		s += nums[i]
		sum[i] = s
	}
	return NumArray{Sum: sum}
}

func (this *NumArray) SumRange(i int, j int) int {
	b := 0
	if i > 0 {
		b = this.Sum[i-1]
	}
	return this.Sum[j] - b
}

func main() {
	fmt.Println("a")
}

File: R/303-range-sum-query-immutable/main.rs
struct Solution {}

struct NumArray {
  s: Vec<i32>,
}

impl NumArray {
  fn new(nums: Vec<i32>) -> Self {
    let mut sum: i32 = 0;
    let mut s: Vec<i32> = Vec::new();
    nums.iter().for_each(|&v| {
      sum += v;
      s.push(sum);
    });
    NumArray { s }
  }

  fn sum_range(&self, left: i32, right: i32) -> i32 {
    if left == 0 {
      self.s[right as usize]
    } else {
      self.s[right as usize] - self.s[left as usize - 1]
    }
  }
}
